### PR TITLE
Tweaks to datetime optimization

### PR DESCRIPTION
**Pull request by rzats**
**Prerequisites**:

- [X] Unless it is a documentation hotfix it should be merged against the `dev` branch
- [X] Branch is up-to-date with the branch to be merged with, i.e. `dev`
- [x] Build is successful
- [X] Code is cleaned up and formatted

### Summary

A followup PR to https://github.com/cmu-delphi/delphi-epidata/pull/977, this makes a few changes to the new datetime optimization code:
- merges `days_to_ranges` and `weeks_to_ranges` into a single `_to_ranges` function with `Callable` arguments
- adds extra tests for some edge cases, as well as `TimePair` objects
- other minor changes

### CODE REVIEW

File: src/server/utils/dates.py [hunk 1]
```diff
@@ -88,63 +89,33 @@ def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) ->
...
-        if m[0] == m[1]:
-            ranges.append(date_to_time_value(m[0]))
-        else:
-            ranges.append((date_to_time_value(m[0]), date_to_time_value(m[1])))
-
-    return ranges
+    return _to_ranges(values, time_value_to_date, date_to_time_value, lambda x: x - timedelta(days=1))
 
 def weeks_to_ranges(values: Sequence[Union[Tuple[int, int], int]]) -> Sequence[Union[Tuple[int, int], int]]:
+    return _to_ranges(values, week_value_to_week, week_to_time_value, lambda x: x - 1)
```
Comment (dshemetov) [inline src/server/utils/dates.py:106]: Can this decrement function correctly handle the first week of they year, like `202201`?

Comment (melange396) [reply src/server/utils/dates.py:106]: if x is a `Week` object, then it should

File: src/server/utils/dates.py [hunk 2]
```diff
@@ -88,63 +89,33 @@ def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) ->
...
+
+def _to_ranges(values: Sequence[Union[Tuple[int, int], int]], value_to_date: Callable, date_to_value: Callable, decrement: Callable) -> Sequence[Union[Tuple[int, int], int]]:
```
Comment (melange396) [inline src/server/utils/dates.py:108]:
```suggestion
def _to_ranges(values: Sequence[Union[Tuple[int, int], int]], value_to_date: Callable, date_to_value: Callable, time_unit: Union[int, timedelta]) -> Sequence[Union[Tuple[int, int], int]]:
```

the idea is to put as much common code into `_to_ranges` as possible, and to simplify the methods that call it as much as possible.  putting lambdas in the one-liners takes away from readability, and specifying just the unit to subtract is clearer.

File: src/server/utils/dates.py [hunk 3]
```diff
@@ -153,18 +124,18 @@ def weeks_to_ranges(values: Sequence[Union[Tuple[int, int], int]]) -> Sequence[U
     for interval in intervals:
         # no overlap, append the interval
         # caveat: we subtract 1 from interval[0] so that contiguous intervals are considered overlapping. i.e. [1, 1], [2, 2] -> [1, 2]
-        if not merged or merged[-1][1] < interval[0] - 1:
+        if not merged or merged[-1][1] < decrement(interval[0]):
```
Comment (melange396) [inline src/server/utils/dates.py:127]:
```suggestion
        if not merged or merged[-1][1] < interval[0] - time_unit:
```

(continuation of suggestion above)

File: src/server/utils/dates.py [hunk 4]
```diff
@@ -88,63 +89,33 @@ def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) ->
...
-            intervals.append([week_value_to_week(v), week_value_to_week(v)])
+            # 20200101 -> [20200101, 20200101]
+            intervals.append([value_to_date(v), value_to_date(v)])
         else: # tuple
-            # (202001, 202002) -> [202001, 202002]
-            intervals.append([week_value_to_week(v[0]), week_value_to_week(v[1])])
+            # (20200101, 20200102) -> [20200101, 20200102]
+            intervals.append([value_to_date(v[0]), value_to_date(v[1])])
 
     intervals.sort(key=lambda x: x[0])
```
Comment (melange396) [inline src/server/utils/dates.py:120]:
```suggestion
    intervals.sort()
```

providing the sort key here is an unnecessary complication which can confuse the reader.

File: src/server/_query.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
 from ._exceptions import DatabaseErrorException
 from ._validate import DateRange, extract_strings
 from ._params import GeoPair, SourceSignalPair, TimePair
-from .utils import dates_to_ranges
+from .utils import dates_to_ranges, days_to_ranges, weeks_to_ranges
```
[the remark this answers is not in the document]
Comment (melange396) [reply src/server/_query.py:25]: also, if we arent going to use the other two methods here, they should come out of the import statement.

**Review by melange396 — CHANGES_REQUESTED** [review]
lookin good!  im sorry if it seems like im picking at nits, but this is a nucleation point for changes that sweep through this call stack, and we will be able to make a lot of changes for the better soon!

i also have a [new comment](https://github.com/cmu-delphi/delphi-epidata/pull/977/files#r986447003) regarding changes to `src/server/_query.py` in the previous PR.

File: src/server/utils/dates.py (hunk 1 above)
Comment (rzats) [reply src/server/utils/dates.py:106]: Yes, the "-" here is a logical operation in the `epiweeks` package, not naive integer subtraction - see https://epiweeks.readthedocs.io/en/stable/usage.html#rich-comparison-and-logical-operations

File: src/server/utils/dates.py [hunk 6]
```diff
@@ -25,11 +26,11 @@ def week_value_to_week(value: int) -> Week:
     return Week(year=year, week=week)
 
 def guess_time_value_is_day(value: int) -> bool:
-    # YYYYMMDD type and not YYYYMM
+    # YYYYMMDD or YYYY-MM-DD type and not YYYYMM
```
[the remark this answers is not in the document]
Comment (rzats) [reply src/server/utils/dates.py:29]: Good point, and changing it to `== 8` doesn't break any tests. I'll make that change

File: src/server/utils/dates.py (hunk 4 above)
Comment (rzats) [reply src/server/utils/dates.py:120]: Looked it up and it's indeed the default behavior, good catch!

**Comment by melange396** [comment]
good stuff, theres just the comment i left in the previous PR about not guessing at the date type when we dont have to: https://github.com/cmu-delphi/delphi-epidata/pull/977/files#r986447003

File: src/server/utils/dates.py [hunk 7]
```diff
@@ -88,83 +89,56 @@ def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) ->
...
+        # convert intervals from dates/weeks back to integers
+        ranges = []
+        for m in merged:
+            if m[0] == m[1]:
+                ranges.append(date_to_value(m[0]))
+            else:
+                ranges.append((date_to_value(m[0]), date_to_value(m[1])))
+
+        return ranges
+    except:
```
Comment (melange396) [inline src/server/utils/dates.py:143]: a `logger.info` that reports the reason for this fall-through would be good.  itll almost certainly be a bad conversion in `value_to_date()`, in which case we might wanna see what bad inputs we are being handed.

File: src/server/utils/dates.py [hunk 8]
```diff
@@ -88,83 +89,56 @@ def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) ->
 
     # determine whether the list is of days (YYYYMMDD) or weeks (YYYYWW) based on first element
     try:
```
Comment (melange396) [inline src/server/utils/dates.py:91]: can we get rid of this `try/except` now too?

**Review by melange396 — CHANGES_REQUESTED** [review]
quick little things and i think itll be good to go!

File: src/server/utils/dates.py (hunk 1 above)
Comment (dshemetov) [reply src/server/utils/dates.py]: The type hints are misleading here then? `values: Sequence[Union[Tuple[int, int], int]]`

File: src/server/utils/dates.py [hunk 9]
```diff
@@ -87,84 +88,55 @@ def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) ->
...
+        ranges = []
+        for m in merged:
+            if m[0] == m[1]:
+                ranges.append(date_to_value(m[0]))
+            else:
+                ranges.append((date_to_value(m[0]), date_to_value(m[1])))
+
+        return ranges
+    except:
+        logging.info('bad input to date ranges', detail=values)
```
Comment (melange396) [inline src/server/utils/dates.py:141]: we should add the actual exception and its message to this too

**Review by melange396 — APPROVED** [review]
